File: app/src/main/java/ui/adapter/FilterCategoryAdapter.java
package ui.adapter;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.AsyncListDiffer;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;
import java.util.List;

import hlv.cute.todo.databinding.ItemCategoryFilterBinding;
import model.Category;

public class FilterCategoryAdapter extends RecyclerView.Adapter<FilterCategoryAdapter.ViewHolder> {

    private final Context context;
    private final AsyncListDiffer<Category> differ;

    public FilterCategoryAdapter(Context context) {
        this.context = context;

        DiffUtil.ItemCallback<Category> diffCallback = new DiffUtil.ItemCallback<Category>() {
            @Override
            public boolean areItemsTheSame(@NonNull Category oldItem, @NonNull Category newItem) {
                return oldItem.getId() == newItem.getId();
            }

            @Override
            public boolean areContentsTheSame(@NonNull Category oldItem, @NonNull Category newItem) {
                return oldItem.compareTo(newItem) == 0;
            }
        };

        differ = new AsyncListDiffer<>(this, diffCallback);
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        ItemCategoryFilterBinding binding = ItemCategoryFilterBinding.inflate(LayoutInflater.from(context), parent, false);
        return new ViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        if (differ.getCurrentList().get(position) != null)
            holder.bind(differ.getCurrentList().get(position));
    }

    @Override
    public int getItemCount() {
        return differ.getCurrentList().size();
    }

    public List<Category> getSelectedItems() {
        List<Category> selected = new ArrayList<>();

        for (int i = 0; i < differ.getCurrentList().size(); i++) {
            Category temp = differ.getCurrentList().get(i);

            if (temp.isSelectedForFilter())
                selected.add(temp);
        }

        return selected;
    }

    static class ViewHolder extends RecyclerView.ViewHolder {
        private final ItemCategoryFilterBinding binding;

        public ViewHolder(@NonNull ItemCategoryFilterBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }

        private void bind(Category category) {
            if (category == null)
                return;

            binding.chip.setText(category.getName());

            binding.chip.setOnCheckedChangeListener(null);
            binding.chip.setChecked(false);
            binding.chip.setOnCheckedChangeListener((buttonView, isChecked) -> {
                if (buttonView.isPressed()) {
                    category.setSelectedForFilter(isChecked);
                }
            });
        }
    }

    public AsyncListDiffer<Category> getDiffer() {
        return differ;
    }
}

File: app/src/main/java/ui/adapter/FilterCategoryAdapter.kt
package ui.adapter

import android.content.Context
import android.view.LayoutInflater
import android.view.ViewGroup
import android.widget.CompoundButton
import androidx.recyclerview.widget.AsyncListDiffer
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import hlv.cute.todo.databinding.ItemCategoryFilterBinding
import model.Category

class FilterCategoryAdapter(private val context: Context) :
    RecyclerView.Adapter<FilterCategoryAdapter.ViewHolder>() {

    val differ: AsyncListDiffer<Category>

    init {
        val diffCallback: DiffUtil.ItemCallback<Category> =
            object : DiffUtil.ItemCallback<Category>() {
                override fun areItemsTheSame(oldItem: Category, newItem: Category): Boolean {
                    return oldItem.id == newItem.id
                }

                override fun areContentsTheSame(oldItem: Category, newItem: Category): Boolean {
                    return oldItem.compareTo(newItem) == 0
                }
            }

        differ = AsyncListDiffer(this, diffCallback)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val binding = ItemCategoryFilterBinding.inflate(
            LayoutInflater.from(
                context
            ), parent, false
        )
        return ViewHolder(binding)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        if (differ.currentList[position] != null) holder.bind(differ.currentList[position])
    }

    override fun getItemCount(): Int {
        return differ.currentList.size
    }

    val selectedItems: List<Category?>
        get() {
            val selected: MutableList<Category> = ArrayList()
            for (i in differ.currentList.indices) {
                val temp = differ.currentList[i]

                if (temp?.isSelectedForFilter == true)
                    selected.add(temp)
            }

            return selected
        }

    inner class ViewHolder(private val binding: ItemCategoryFilterBinding) :
        RecyclerView.ViewHolder(
            binding.root
        ) {

        fun bind(category: Category?) {
            if (category == null)
                return

            binding.root.requestLayout()

            binding.chip.requestLayout()

            binding.chip.isChecked = category.isSelectedForFilter

            with(category) {
                if (id == 0 && name == null)
                    binding.chip.text = "-بدون دسته‌بندی-"
                else
                    binding.chip.text = name
            }


            binding.chip.setOnCheckedChangeListener(null)

            binding.chip.setOnCheckedChangeListener { buttonView: CompoundButton, isChecked: Boolean ->
                if (buttonView.isPressed)
                    category.isSelectedForFilter = isChecked
            }
        }
    }
}